### ISSUE TITLE
More flexibility in array-element-newline

### ISSUE THREAD

**Issue by LFP6** [eslint/eslint#9054]
I would like to propose some additional options for the new array-element-newline rule. The existing options don't feel flexible enough for me. There are a couple of different thoughts I had in mind (note, I have a number of thoughts here to help brainstorm - I doubt all of them would be useful in practice). Also please excuse any sloppy/awkward naming/syntax, this is still relatively new for me (criticism welcome to help clarify this request).

* `"expressions-only"`, `"with-expressions"` and `{"break": <"before"|"after"|"both">}`: In addition to "always" and "never", I would like to specify an option to specify that when an expression is encountered in an array, it should trigger a newline. For "expressions-only", literals would still be able to be placed on the same line, but the expression would be delegated to its own line, whereas "with-expressions" would cause all items in the array to be on their own lines. The "break" option would allow whether to create a newline before, after, or before and after the expression, for the situation of wanting `},{` vs `},\n{` betweem objects in an array. An example of where I would use this:
```js
class Person {
    constructor(firstName, lastName, email) { ... }
}

// I like this, it's concise
let firstNames = ['Bob', 'Sam', 'Fred']

// This is not ok to me, as the individual elements are unreadable
let people = [
    new Person('Bob', 'Smith', 'bsmith@example.com'), new Person('Sam', 'Smith', 'ssmith@example.com'), new Person('Fred', 'Smith', 'fsmith@example.com'),
]

// This would be preferred
let people = [
    new Person('Bob', 'Smith', 'bsmith@example.com'),
    new Person('Sam', 'Smith', 'ssmith@example.com'),
    new Person('Fred', 'Smith', 'fsmith@example.com'),
]
```
* `{"arraysInArrays": <"always"|"before"|"after"|"beforeAndAfter"|"never">}`
* `{"objectsInArrays": <"always"|"before"|"after"|"beforeAndAfter"|"never">}`
* `{"functionDefinitionsInArrays": <"always"|"before"|"after"|"beforeAndAfter"|"never">}`
* `{"functionCallsInArrays": <"always"|"before"|"after"|"beforeAndAfter"|"never">}`
I'm still new to eslint, so I don't completely know the right way to handle this (perhaps allowing `<any AST type>InArrays`?). The idea is that you could further specify if you wanted to enable/disable/change the break on a specific type of expression.
* `{"maxLineCharLength": <number>}`: [max-len](http://eslint.org/docs/rules/max-len) exists, so I imagine it should be possible to split on commas if the line gets to be too long (and for this reason, should override/extend preferences like the one above). IE:
```js
// Bad with {"maxCharLength": 120}
let arr = ["verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere"]

// Good
let arr = ["verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere",
"verylongitemgoeshere", "verylongitemgoeshere", "verylongitemgoeshere"]

// Or maybe there needs to be a way to specify something like the indenting here?
let arr = ["verylongitemgoeshere", "reallylongitemgoeshere", "anotherlongitemgoeshere", "verylongitemgoeshere",
           "somelongitemgoeshere", "alongitemgoeshere", "verylongitemgoeshere"]
```
* `{"maxElementLength": <number>}`: This would enforce line breaks when there are more elements in the array than the specified number. I originally thought I wanted this, but I think the expressions and maxCharLength solutions are what I'm really looking for. Wondering someone would still find this useful.
* `{"consistantColumns": <boolean>}`: A convention I've seen before is having an equal number of array elements per column, across multiple lines. Perhaps an exclusion to [no-multi-spaces](http://eslint.org/docs/rules/no-multi-spaces) needs to be added as well? Example:
```js
let arr = [
    "First", "Second", "Third",
    "Third", "Fourth", "Fifth"
]

// Some people like their columns to line up regardless of length - also, how might this play with max-len?
let arr = [
    "First item", "Second", "Third",
    "Third",      "Fourth", "Fifth"
]
```

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. I have a few initial thoughts:

> "expressions-only"

What do you mean by an "expression" here? Technically, all array elements are expressions (they are nodes in the syntax tree which will have computed values at runtime).

> arraysInArrays, objectsInArrays, functionDefinitionsInArrays, functionCallsInArrays

I think having a [selector](http://eslint.org/docs/developer-guide/selector) option would be better than having to implement individual options for each node type.

> maxLineCharLength

This has been a proposed in a few forms before (e.g. https://github.com/eslint/eslint/issues/7526). The issue is that the line length of an array depends on a number of other stylistic choices (e.g. comma spacing, spacing within the elements) and the `array-element-newline` rule can't read that configuration.

> consistentColumns

I think this is out of scope for the `array-element-newline` rule, since it only deals with newlines between elements and not other spacing.

**Comment by LFP6** [comment]
All good thoughts, and making me think a bit more about what I'm really looking for.

> What do you mean by an "expression" here?

Generally, I'm just referring to "not a literal" (`'string literal'`, `5`, etc). As I'm thinking about it, I'm not 100% sure if this is what I want - maybe I'm looking for character length of each element, ie:
```js
// Bad
let arr = ['first really long element goes over here', 'another long element going over there', 'a third lengthy element in this spot']

// Good
let arr2 = [
    'first really long element goes over here',
    'another long element going over there',
    'a third lengthy element in this spot'
]

// Also fine
let arr3 = [3, 4, 5]
```
I had thought that in this case I'd still want the strings on one line, but I'm reconsidering - forgot to include this in the initial description. I imagine this would wind up being more useful, and also clearer.

> I think having a [selector](http://eslint.org/docs/developer-guide/selectors) option would be better than having to implement individual options for each node type.

Yeah. Maybe do something like `{"always-type": [<type>[, <type>]]}` or so? Admittedly this parameter is somewhat complex, and not entirely relevant if element char length is used instead of element type.

> This has been a proposed in a few forms before (e.g. #7526). The issue is that the line length of an array depends on a number of other stylistic choices (e.g. comma spacing, spacing within the elements) and the array-element-newline rule can't read that configuration.

I was actually thinking about this in the opposite direction - forcing it to be split across multiple lines if it's too long, but I didn't consider how you would go about determining if it is incorrectly placed on multiple lines. I could see the possibility of exposing the "same" configuration as a property (so, `{"maxLineCharLength": {"length": <number>, "commaSpacing": [<config>]` or something, but that does admittedly get a bit messy. I do like that rule suggestion a lot as well, actually - I'd really love to see a way to make it work since I'd consider that a relatively noticeable style issue. This is the best way I can think of getting around the state issue.

> I think this is out of scope for the array-element-newline rule, since it only deals with newlines between elements and not other spacing.

Yep. Not something I'd personally use anyways, but if someone is looking for that it could potentially be a new rule proposal.

**Comment by LFP6** [comment]
Curious if anyone has any thoughts about adding a max char length per element. I'd really like to have the split based on total length, but per-element at least would be a huge help.

**Comment by LFP6** [comment]
As I'm thinking of it. I'm thinking max char length + max number of elements per line could be "good enough" at least. This way, the user is responsible for creating a combination of the two that is less than their max character length - that is, if the largest number of elements had the longest number of characters (including spacing), it would never go beyond the max line length.

```javascript
// If I have this array, 56 characters long, and a 50 character line limit
['0123456789', '0123456789', '0123456789', '0123456789']

// I could set the max length of each element to 9, in which case this array would get split
// according to other rules
['0123456789',
'0123456789',
'0123456789',
'0123456789']

// Otherwise, I could set my max element length to 10, but say only allow 3 elements per line
['0123456789', '0123456789', '0123456789',
'0123456789']
```
Either way, the max characters stays under 50 - except for really long individual elements where on it's own line it's too long, but that gets caught by max-len and would get handled according to relevant rules. Of course, I haven't done the math for the optimal numbers in this case, but that's the general idea.

**Comment by LFP6** [comment]
As I'm thinking, it does get a little complicated, as when assigning it to a variable, you don't know how long that variable is, for example. I'd be interested if there's a better way to handle this, but I can't immediately think of one.

**Comment by not-an-aardvark** [comment]
I'm unsure whether checking the max char length of an element is a good idea, because the character length of an element could be influenced by a number of other stylistic preferences which are handled by other rules. If there were two separate rules switching behavior based on the character length of a piece of code, then it seems like it would be easy to accidentally create a situation where the rules can't both be satisfied. (For example, fixing one rule could require adding characters, then the other rule might change behavior because the character length has changed, and require the characters to be removed again.) Similarly, switching behavior based on character length could sometimes lead to issues where either behavior is correct for particular code, because adding linebreaks would also affect the character length.

For these reasons, we've generally  avoided switching behavior based on character length.

**Comment by LFP6** [comment]
That's an interesting point. I'm curious how many people take the same approach I do for arrays - the couple people I talked to in ##javascript on Freenode when originally thinking about this seemed to use that approach if I'm remembering correctly. All else fails I guess that could be a separate plugin.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
